src/checkers/inference/model/ConstraintManager.java

```diff
@@ -37,171 +30,186 @@
+     */
     public PreferenceConstraint createPreferenceConstraint(VariableSlot variable, ConstantSlot goal,
             int weight) {
-        if (variable == null || goal == null) {
-            ErrorReporter.errorAbort("Create preference constraint with null argument. Variable: "
-                    + variable + " Goal: " + goal);
-        }
-        return new PreferenceConstraint(variable, goal, weight, getCurrentLocation());
+        return PreferenceConstraint.create(variable, goal, weight, getCurrentLocation());
     }
 
+    /**
+     * Creates an {@link ExistentialConstraint} for the given slot and lists of constraints.
+     */
     public ExistentialConstraint createExistentialConstraint(Slot slot,
             List<Constraint> ifExistsConstraints, List<Constraint> ifNotExistsConstraints) {
-        // TODO: add null checking for argument.
-        return new ExistentialConstraint((VariableSlot) slot,
-                ifExistsConstraints, ifNotExistsConstraints, getCurrentLocation());
+        return ExistentialConstraint.create((VariableSlot) slot, ifExistsConstraints,
+                ifNotExistsConstraints, getCurrentLocation());
     }
 
+    // TODO: give location directly in Constraint.create() methods
     private AnnotationLocation getCurrentLocation() {
         if (visitorState.getPath() != null) {
-            return VariableAnnotator.treeToLocation(inferenceTypeFactory, visitorState.getPath()
-                    .getLeaf());
+            return VariableAnnotator.treeToLocation(inferenceTypeFactory,
+                    visitorState.getPath().getLeaf());
         } else {
             return AnnotationLocation.MISSING_LOCATION;
         }
     }
 
+    // All addXXXConstraint methods create a (possibly normalized) constraint for the given slots
+    // and issues errors for unsatisfiable constraints.
+
+    /**
+     * Creates and adds a {@link SubtypeConstraint} between the two slots to the constraint set,
+     * which may be normalized to {@link AlwaysTrueConstraint} or {@link EqualityConstraint}. An
+     * error is issued if the {@link SubtypeConstraint} is always unsatisfiable.
+     */
     public void addSubtypeConstraint(Slot subtype, Slot supertype) {
-        if ((subtype instanceof ConstantSlot)
-                && this.realQualHierarchy.getTopAnnotations().contains(((ConstantSlot) subtype).getValue())) {
-            this.addEqualityConstraint(supertype, (ConstantSlot) subtype);
-        } else if ((supertype instanceof ConstantSlot)
-                && this.realQualHierarchy.getBottomAnnotations().contains(
-                        ((ConstantSlot) supertype).getValue())) {
-            this.addEqualityConstraint(subtype, (ConstantSlot) supertype);
-        } else {
-            this.add(this.createSubtypeConstraint(subtype, supertype));
-        }
+        add(createSubtypeConstraint(subtype, supertype), SubtypeConstraint.class);
```

Comment:
The `add()` method is refactored to take in a class literal to generate an error message key of the form "xxx.constraint.unsatisfiable" if the normalized constraint is `AlwaysFalseConstraint`.

There may be some use for tracking the original constraint (as a class literal) that some normalized constraint was derived from, so that we can look it up for creating the error message key in `add()`. That would reduce the verbosity here and below, and eliminate the chance that we somehow pass in an incorrect literal. The implementation would require adding a class literal parameter to each `create()` method in the constants. For now it doesn't seem like we will create any normalization chains (ie A normalizes to B, B then normalizes to C). If we do, we will need to propagate the original constraint literal down to all child constraints.

@wmdietl @CharlesZ-Chen any thoughts on whether we should add tracking into the constraints?

Reply:
I think adding tracking maybe better than requiring a class literal parameter. Also it maybe benefit to keep tracking of original constraint of a normalized result.

Reply:
What useful information would we get by keeping the object of the original constraint? As currently implemented, we can get the slots for an enhanced error message from the parameters of `addXXXConstraint`, thus they don't have to be stored in the normalized constraint.

The alternative design is that `create()` and `add/createXXXConstraint` does not perform normalization, but we add a `public Constraint normalize()` to every Constraint, and call this within `ConstraintManager.add()` so that only normalized constraints are added to the set. If there are errors to be issued, we have the precise original constraint type available in `add()` and also the normalized constraint to use.

`InferenceQualHier` can also call normalize to see if the subtype relation between two slots is always false, satisfying 136.